src/vsr/replica.zig

```diff
@@ -3970,6 +4101,49 @@ pub fn ReplicaType(
             return @sizeOf(vsr.ReconfigurationResult);
         }
 
+        fn commit_upgrade(
+            self: *Self,
+            prepare: *const Message.Prepare,
+            output_buffer: *align(16) [constants.message_body_size_max]u8,
+        ) usize {
+            maybe(self.upgrade_release == null);
+            assert(self.commit_stage == .setup_client_replies);
+            assert(self.commit_prepare.? == prepare);
+            assert(self.superblock.working.vsr_state.checkpoint.release == self.release);
+            assert(prepare.header.command == .prepare);
+            assert(prepare.header.operation == .upgrade);
+            assert(prepare.header.size == @sizeOf(vsr.Header) + @sizeOf(vsr.UpgradeRequest));
+            assert(prepare.header.op == self.commit_min + 1);
+            assert(prepare.header.op <= self.op);
+            assert(prepare.header.client == 0);
+
+            if (self.pipeline == .queue) {
+                assert(self.pipeline.queue.prepare_queue.count == 1);
+                assert(self.pipeline.queue.request_queue.empty());
+            }
+
+            const request = std.mem.bytesAsValue(
+                vsr.UpgradeRequest,
+                prepare.body()[0..@sizeOf(vsr.UpgradeRequest)],
+            );
+            assert(request.release >= self.release);
+
+            if (request.release == self.release) {
+                // We are replaying this upgrade request after restarting into the new version.
+                assert(self.upgrade_release == null);
+            } else {
+                if (self.upgrade_release) |upgrade_release| {
+                    assert(upgrade_release == request.release);
```

Comment:
Once an "upgrade to release X" is committed, the replica cannot change its mind and decide to commit "upgrade to release Y" (for any Y≠X) until release X has checkpointed and the replica is running release X.

(We could ease this restriction later if necessary, but I think it is a helpful simplification for my mental model if nothing else.)